internal/impl/opensearch/output.go

```diff
@@ -0,0 +1,454 @@
+				boff.Reset()
+			}
+			return boff.NextBackOff()
+		},
+		MaxRetries: int(e.conf.MaxRetries),
+	}
+	if e.conf.Auth.Enabled {
+		opts.Username = e.conf.Auth.Username
+		opts.Password = e.conf.Auth.Password
+	}
+
+	opts.Transport = &http.Transport{
+		ResponseHeaderTimeout: e.timeout,
+	}
+	if e.conf.TLS.Enabled {
+		opts.Transport = &http.Transport{
+			TLSClientConfig:       e.tlsConf,
+			ResponseHeaderTimeout: e.timeout,
+		}
+
+	}
+
+	awsOpts, err := AWSOptFn(opts.Transport, e.conf)
+	if err != nil {
+		return err
+	}
+	if awsOpts != nil {
+		opts.Transport = awsOpts
+
+	}
+
+	if e.conf.GzipCompression {
+		opts.CompressRequestBody = true
+	}
+
+	client, err := os.NewClient(opts)
+	if err != nil {
+		return err
+	}
+
+	e.client = client
+	e.log.Infof("Sending messages to OpenSearch index at urls: %s\n", e.urls)
+	return nil
+}
+
+type pendingBulkIndex struct {
+	Action   string
+	Index    string
+	Pipeline string
+	Routing  string
+	Type     string
+	Doc      interface{}
+	ID       string
+}
+
+// WriteBatch will attempt to write a message to OpenSearch, wait for
+// acknowledgement, and returns an error if applicable.
+func (e *OpenSearch) WriteBatch(ctx context.Context, msg message.Batch) error {
+	return e.Write(msg)
+}
+
+// Write will attempt to write a message to OpenSearch, wait for
+// acknowledgement, and returns an error if applicable.
+func (e *OpenSearch) Write(msg message.Batch) error {
+	if e.client == nil {
+		return component.ErrNotConnected
+	}
+
+	requests := make([]*pendingBulkIndex, msg.Len())
+	if err := msg.Iter(func(i int, part *message.Part) error {
+		jObj, ierr := part.AsStructured()
+		if ierr != nil {
+			e.log.Errorf("Failed to marshal message into JSON document: %v\n", ierr)
+			return fmt.Errorf("failed to marshal message into JSON document: %w", ierr)
+		}
+		requests[i] = &pendingBulkIndex{
+			Action:   e.actionStr.String(i, msg),
+			Index:    e.indexStr.String(i, msg),
+			Pipeline: e.pipelineStr.String(i, msg),
+			Routing:  e.routingStr.String(i, msg),
+			Type:     e.typeStr.String(i, msg),
+			Doc:      jObj,
+			ID:       e.idStr.String(i, msg),
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+	start := time.Now()
+	b, _ := opensearchutil.NewBulkIndexer(opensearchutil.BulkIndexerConfig{
+		FlushInterval: time.Second * 1,
+		FlushBytes:    0,
+		Client:        e.client,
+	})
+	ctx := context.Background()
+	for _, v := range requests {
+		bulkReq, err := e.buildBulkableRequest(v)
+		if err != nil {
+			return err
+		}
+		_ = b.Add(ctx, *bulkReq)
+	}
+
+	_ = b.Close(ctx)
+
+	biStats := b.Stats()
+	dur := time.Since(start)
+
+	if biStats.NumFailed > 0 {
+		e.log.Errorf(
+			"Indexed [%s] documents with [%s] errors in %s (%s docs/sec)",
+			humanize.Comma(int64(biStats.NumFlushed)),
+			humanize.Comma(int64(biStats.NumFailed)),
+			dur.Truncate(time.Millisecond),
+			humanize.Comma(int64(1000.0/float64(dur/time.Millisecond)*float64(biStats.NumFlushed))),
+		)
+	} else {
+		e.log.Infof(
```

Comment:
This might get noisy, I think. Might be better to use `Debugf`.

Reply:
Done! I love these entire review feedback of nitty-gritty details